js/ui/payg.js

```diff
@@ -441,6 +441,65 @@ function timeToString(seconds) {
     return ("%s %s").format(daysStr, hoursStr);
 }
 
+//In cases the gnome-shell hangs up more than 1 sec, we want to show
+//the exact time inputted. This assumes that the flow will never hang up
+//more than the minimum amount available for credit, which is 5 sec.
+function roundSecs(seconds) {
+    seconds = Math.ceil(seconds);
+
+    if (seconds % 5 > 0)
+        seconds -= (seconds % 5) + 5;
+
+    return seconds;
+}
+
+//Similar to timeToStrings, but does not process partial time,
+//since it's meant to be used for expiration time changes, with
+//a wider range of periods
+function successMessage(seconds) {
+
+    seconds = roundSecs(seconds);
+    let messageStr = null;
+
+    if (seconds < 60) {
+        messageStr = Gettext.ngettext("%s second has", "%s seconds have", seconds).format(Math.floor(seconds));
+        return _("%s been added to your %s credit.").format(messageStr, NOTIFICATION_TITLE_TEXT);
+    }
+
+    let minutes = Math.floor(seconds / 60);
+    if (minutes < 120) {
+        messageStr = Gettext.ngettext("%s minute has", "%s minutes have", minutes).format(minutes);
+        return _("%s been added to your %s credit.").format(messageStr, NOTIFICATION_TITLE_TEXT);
+    }
+
+    let hours = Math.floor(minutes / 60);
+    if (hours < 24) {
+        messageStr = Gettext.ngettext("%s hour has", "%s hours have", hours).format(hours);
+        return _("%s been added to your %s credit.").format(messageStr, NOTIFICATION_TITLE_TEXT);
+    }
+
+    let days = Math.floor(hours / 24);
+    if (days < 30) {
```

Comment:
This has kinda passed by me, but for the success message, the user will rarely buy a product with ""broken"" time, right? Based on that, I put the minutes to show `minutes < 60`, so when buying an hour of credit, it shows 1 Hour, instead of 60 minutes ( basing on the assumption it's more a more common product wording ).

Reply:
The reason I prefer `threshold * 2` (eg, 120 minutes) is to compensate for the inaccuracy of using `floor()`.

Eg, if someone has 119 minutes left, they'd probably much prefer to see that than "1 hour" which is basically half the remaining time. Same for 47 hours vs. 1 day. It'd be more-accurate to simply round but we can't ever round up because then we'd tell the user they could use the computer longer than they actually can. And that would matter if, eg, they're a university student whose open-computer CS exam ends in 47 hours but they only have 40 hours of credit left (and we told them they have >= 48 hours since we rounded to "2 days"). Better to tell them they have 40 hours in this case or 2 days (in case they have 2.7 days left) and then lock at a slightly later time. (And we'll still give them more accurate times as they approach the actual lock time since we switch to a finer-grained time scale).